lib/neo4j/active_node/query/query_proxy_eager_loading.rb

```diff
@@ -2,58 +2,153 @@ module Neo4j
+          end
+
+          def add_association(spec)
+            if spec.is_a?(Array)
+              spec.each { |s| add_association(s) }
+            elsif spec.is_a?(Hash)
+              spec.each { |k, v| (self[k] ||= AssociationTree.new(model, k)).add_association(v) }
+            else
+              self[spec] ||= AssociationTree.new(model, spec)
             end
+          end
+
+          def paths(*prefix)
+            values.flat_map { |v| [[*prefix, v]] + v.paths(*prefix, v) }
+          end
 
-            yield(record)
+          private
+
+          def target_class(model, key)
+            association = model.associations[key]
+            fail "Invalid association: #{[*path, key].join('.')}" unless association
+            model.associations[key].target_class
           end
         end
 
-        def with_associations_spec
-          @with_associations_spec ||= []
+        def pluck_vars(node, rel)
+          return super if with_associations_spec.size.zero?
+
+          perform_query
         end
 
-        def with_associations(*spec)
-          invalid_association_names = spec.reject do |association_name|
-            model.associations[association_name]
-          end
+        def perform_query
+          @_cache = IdentityMap.new
+          query_from_association_spec
+            .map do |record, eager_data|
+            cache_and_init(record, with_associations_spec)
+            eager_data.zip(with_associations_spec.paths.map(&:last)).each do |eager_records, element|
+              eager_records.first.zip(eager_records.last).each do |eager_record|
+                add_to_cache(*eager_record, element)
+              end
+            end
 
-          if !invalid_association_names.empty?
-            fail "Invalid associations: #{invalid_association_names.join(', ')}"
+            record
           end
+        end
 
+        def with_associations(*spec)
           new_link.tap do |new_query_proxy|
-            new_spec = new_query_proxy.with_associations_spec + spec
-            new_query_proxy.with_associations_spec.replace(new_spec)
+            new_query_proxy.with_associations_spec = with_associations_spec.clone
+            new_query_proxy.with_associations_spec.add_association(spec)
           end
         end
 
+        def with_associations_spec
+          @with_associations_spec ||= AssociationTree.new(model)
+        end
+
+        def with_associations_spec=(tree)
+          @with_associations_spec = tree
+        end
+
         private
 
-        def with_associations_return_clause(variables = with_associations_spec)
-          variables.map { |n| "#{n}_collection" }.join(',')
+        def add_to_cache(rel, node, element)
+          direction = element.association.direction
+          if rel.is_a?(Neo4j::ActiveRel)
+            rel.instance_variable_set(direction == :in ? '@from_node' : '@to_node', node)
+          end
+          @_cache[direction == :out ? rel.start_node_neo_id : rel.end_node_neo_id]
```

Comment:
In testing this out, I got an error that `start_node_neo_id` doesn't exist.  This means that the `gemspec` for this change would need to be changed to require `7.2.2` of the `neo4j-core` gem.  Probably we can release as 8.2.x of the `neo4j` gem, so that shouldn't be a big deal